src/CairoLib.sol

```diff
@@ -205,4 +254,60 @@ library CairoLib {
         (bool success,) = CAIRO_MESSAGING_ADDRESS.call(messageData);
         require(success, "CairoLib: sendMessageToL1 failed");
     }
+
+    /// @notice Converts a Cairo ByteArray to a string
+    /// @dev A ByteArray is represented as:
+    /// 1. An array of fully packed 31-byte words (data)
+    /// 2. A pending word (felt252) with < 31 bytes
+    /// 3. The length of the pending word (0-30 bytes)
+    /// @param data An Cairo representation of the ByteArray, serialized to individual bytes.
+    function byteArrayToString(bytes memory data) internal view returns (string memory) {
+        require(data.length >= 96, "Invalid byte array length");
+
+        uint256 fullWordsLength;
+        uint256 fullWordsPtr;
+        uint256 pendingWord;
+        uint256 pendingWordLen;
+
+        assembly {
+            fullWordsLength := mload(add(data, 32))
+            let fullWordsByteLength := mul(fullWordsLength, 32)
+            fullWordsPtr := add(data, 64)
+            let pendingWordPtr := add(fullWordsPtr, fullWordsByteLength)
+            pendingWord := mload(pendingWordPtr)
+            pendingWordLen := mload(add(pendingWordPtr, 32))
+        }
+
+
+        require(pendingWordLen <= 31, "Invalid pending word length");
+
+        uint256 totalLength = fullWordsLength * 31 + pendingWordLen;
+        bytes memory result = new bytes(totalLength);
+        uint256 resultPtr;
+
+        assembly {
+            resultPtr := add(result, 32)
+        }
+
+        // Copy full words. Because of the Cairo -> Solidity conversion,
+        // each full word is 32 bytes long, but contains 31 bytes of information.
+        for (uint256 i = 0; i < fullWordsLength; i++) {
+            assembly {
+                let word := mload(fullWordsPtr)
+                let storedWord := shl(8, word)
+                mstore(resultPtr, storedWord)
+                resultPtr := add(resultPtr, 31)
+                fullWordsPtr := add(fullWordsPtr, 32)
+            }
+        }
+
+        // Copy pending word
+        if (pendingWordLen > 0) {
+            assembly {
+                mstore(resultPtr, shl(mul(sub(32, pendingWordLen), 8), pendingWord))
+            }
+        }
```

Comment:
You could do  the last 3 block in one assembly block
It would avoid context switch imo
```
assembly {
    resultPtr := add(result, 32)
    for { let i := 0 } lt(i, fullWordsLength) { i := add(i, 0x20) } {
        ...
    }
    if iszero(eq(pendingWordLen, 0)) {
        mstore(resultPtr, shl(mul(sub(32, pendingWordLen), 8), pendingWord))
   }
}

```
(To test)